File: frontend/src/lib/auth/passwordPolicy.ts
/**
 * Advisory mirror of `backend/src/domain/password_policy.py`.
 *
 * This is UX-only live feedback for the `/setup` form. `POST /auth/setup` is
 * the authoritative validator; drift between the two is safe because the
 * server always re-validates and never trusts this client-side check.
 */

export const MIN_LENGTH = 12;

export type PasswordRule = "min_length" | "uppercase" | "lowercase" | "digit" | "special";

const UPPERCASE_RE = /[A-Z]/;
const LOWERCASE_RE = /[a-z]/;
const DIGIT_RE = /\d/;
const SPECIAL_RE = /[^A-Za-z0-9]/;

/** Returns every unmet policy rule; an empty array means the password is compliant. */
export function validatePassword(password: string): PasswordRule[] {
  const violations: PasswordRule[] = [];
  if (password.length < MIN_LENGTH) violations.push("min_length");
  if (!UPPERCASE_RE.test(password)) violations.push("uppercase");
  if (!LOWERCASE_RE.test(password)) violations.push("lowercase");
  if (!DIGIT_RE.test(password)) violations.push("digit");
  if (!SPECIAL_RE.test(password)) violations.push("special");
  return violations;
}

File: backend/src/domain/password_policy.py
"""Password policy for dashboard credentials (first-run setup).

Single source of truth: server-side validation always wins over the
frontend's advisory mirror. Returns EVERY unmet rule so the caller can
surface a complete error list in one round trip, instead of one-at-a-time.
"""

import re

MIN_LENGTH = 12

_UPPERCASE_RE = re.compile(r"[A-Z]")
_LOWERCASE_RE = re.compile(r"[a-z]")
_DIGIT_RE = re.compile(r"\d")
_SPECIAL_RE = re.compile(r"[^A-Za-z0-9]")


def validate(password: str) -> list[str]:
    """Return the list of unmet policy rules; empty when compliant.

    Possible rule identifiers: ``min_length``, ``uppercase``, ``lowercase``,
    ``digit``, ``special``.
    """
    violations: list[str] = []
    if len(password) < MIN_LENGTH:
        violations.append("min_length")
    if not _UPPERCASE_RE.search(password):
        violations.append("uppercase")
    if not _LOWERCASE_RE.search(password):
        violations.append("lowercase")
    if not _DIGIT_RE.search(password):
        violations.append("digit")
    if not _SPECIAL_RE.search(password):
        violations.append("special")
    return violations
